teps/0118-matrix-with-explicit-combinations-of-parameters.md

```diff
@@ -0,0 +1,586 @@
+          - name: s390x-no-race 
+            params:
+              - name: GOARCH
+                value: "linux/s390x"
+              - name: flags
+                value: "-cover -v" 
+    - ...
+```
+
+The `matrix.include` section when specified with `matrix.params` is used to expand existing auto-populated `Matrix`. 
+The above `Matrix` specification will create three `taskRuns`, one for each architecture in `GOARCH` (`"linux/amd64"`, 
+`"linux/ppc64le"`, and `"linux/s390x"`) and with default `flags`. Now, when a `taskRun` is created for the `GOARCH` 
+value of `linux/s390x`, an additional parameter called `flags` with the value of `-cover -v` will be included in the
+`taskRun`. The `matrix.include` section can list `params` which does not exist in the `matrix.params` section. In this 
+example specification, `flags` was not listed in `matrix.params`. At the same time, `matrix.include` section can 
+list `params` which does exist in the `matrix.params` section.
+
+Let's expand this use case to showcase a little more complex combinations, for example:
+
+```yaml
+      matrix:
+        params:
+          - name: GOARCH
+            value:
+              - "linux/amd64"
+              - "linux/ppc64le"
+              - "linux/s390x"
+          - name: version
+            value:
+              - "go1.17"
+              - "go1.18.1"
+```
+
+This `matrix` specification will result in six `taskRuns` with the following `matrix` combinations:
+
+```json!
+{ "GOARCH": "linux/amd64", "version": "go1.17" }
+
+{ "GOARCH": "linux/amd64", "version": "go1.18.1" }
+
+{ "GOARCH": "linux/ppc64le", "version": "go1.17" }
+
+{ "GOARCH": "linux/ppc64le", "version": "go1.18.1" }
+
+{ "GOARCH": "linux/s390x", "version": "go1.17" }
+
+{ "GOARCH": "linux/s390x", "version": "go1.18.1" }
+```
+
+Now, let's introduce `include` with a couple of `Parameters`: `"package"`, `"flags"` and `"context"`:
+
+```yaml=
+      matrix:
+        params:
+          - name: GOARCH
+            value:
+              - "linux/amd64"
+              - "linux/ppc64le"
+              - "linux/s390x"
+          - name: version
+            value:
+              - "go1.17"
+              - "go1.18.1"
+        include:
+          - name: common-package
+            params:
+              - name: package
+                value: "path/to/common/package/"
+          - name: s390x-no-race 
+            params:
+              - name: GOARCH
+                value: "linux/s390x"
+              - name: flags
+                value: "-cover -v" 
+          - name: go117-context
+            params:
+              - name: version
+                value: "go1.17"
+              - name: context
+                value: "path/to/go117/context"
+         - name: non-existent-arch
+           params:
+              - name: GOARCH
+                value: "I-do-not-exist"
+```
+
+The first `include` clause is added to all the original `matrix` combintations without overwriting any `parameters` of 
```

Comment:
I'm curious why someone would do this as opposed to just using it as a regular parameter outside of the matrix.

Reply:
only reason I can imagine is personal preference to have every parameter under matrix when they are fanning out - but the effect is the same as having the param outside the matrix - this example mainly makes it clear what happens in this scenario so that it's not underspecified

note that this is strongly influenced by `include` in `matrix` in GHA that has the same behavior: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrixinclude